source/internal/usb_utils.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package busutils
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+type UsbDeviceInfo map[string]string
+type UsbClassMap map[string]UsbDeviceInfo
+
+var DefaultUsbDevAttrs = []string{"class", "vendor", "device"}
+
+// The USB device sysfs files do not have terribly user friendly names, map
+// these for consistency with the PCI matcher.
+var devAttrFileMap = map[string]string{
+	"class":  "bDeviceClass",
+	"device": "idProduct",
+	"vendor": "idVendor",
+}
+
+func readSingleUsbSysfsAttribute(path string) (string, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to read device attribute %s: %v", filepath.Base(path), err)
+	}
+
+	attrVal := strings.TrimSpace(string(data))
+
+	return attrVal, nil
+}
+
+// Read a single USB device attribute
+// A USB attribute in this context, maps to the corresponding sysfs file
+func readSingleUsbAttribute(devPath string, attrName string) (string, error) {
+	return readSingleUsbSysfsAttribute(path.Join(devPath, devAttrFileMap[attrName]))
+}
+
+// Read information of one USB device
+func readUsbDevInfo(devPath string, deviceAttrSpec map[string]bool) (UsbClassMap, error) {
+	classmap := UsbClassMap{}
+	info := UsbDeviceInfo{}
+
+	for attr, _ := range deviceAttrSpec {
+		attrVal, _ := readSingleUsbAttribute(devPath, attr)
+		if len(attrVal) > 0 {
+			info[attr] = attrVal
+		}
+	}
+
+	// USB devices encode their class information either at the device or the interface level. If the device class
+	// is set, return as-is.
+	if info["class"] != "00" {
+		classmap[info["class"]] = info
+	} else {
+		// Otherwise, if a 00 is presented at the device level, descend to the interface level.
+		interfaces, err := filepath.Glob(devPath + "/*/bInterfaceClass")
+		if err != nil {
+			return classmap, err
+		}
+
+		// A device may, notably, have multiple interfaces with mixed classes, so we create a unique device for each
+		// unique interface class.
+		for _, intf := range interfaces {
+			// Determine the interface class
+			attrVal, err := readSingleUsbSysfsAttribute(intf)
+			if err != nil {
+				return classmap, err
+			}
+
+			// Update interface class
+			info["class"] = attrVal
```

Comment:
Instead of modifying, I think we need to create a new copy of `info` here. Otherwise we end up modifying the same instance of the map.

Reply:
This was intentional, as we copy it on the next line when inserting it into the classmap, which is the thing that is returned from the function. We never use `info` again, so clobbering "class" on each successive iteration is perfectly fine.

Reply:
The problem with this is that with `classmap[attrVal] = info` you don't create a new copy of the map, merely just a reference is passed. Thus, later changes to info are also reflected in other `classmap` members created in this for loop.

I.e. if you do somehting like:
```
	a := map[string]string{"foo": "1", "bar": "2"}
	b := a
	b["foo"] = "123"
```
then `a["foo"] == "123"`

Reply:
Wow, I thought this only applied to pointers, but you're right, the same thing applies to maps. Learn something new every day. I've now replaced this with a local copy for each iteration.

Reply:
👍 Beware that this applies to slices, too 😉